src/groups/mqb/mqba/mqba_authenticator.cpp

```diff
@@ -155,11 +156,348 @@ int Authenticator::sendAuthenticationMessage(
+
+    if (authnRc != 0) {
+        BALL_LOG_ERROR << "Authentication failed for connection '"
+                       << channel->peerUri() << "' with mechanism '"
+                       << request.mechanism() << "' [rc: " << authnRc
+                       << ", error: " << errorStream.str() << "]";
+
+        response->status().code() = authnRc;
+        response->status().category() =
+            bmqp_ctrlmsg::StatusCategory::E_REFUSED;
+        response->status().message() = errorStream.str();
+    }
+    else {
+        response->status().code() = 0;
+        response->status().category() =
+            bmqp_ctrlmsg::StatusCategory::E_SUCCESS;
+        response->lifetimeMs() = result->lifetimeMs();
+
+        authenticationContext->setAuthenticationResult(result);
+
+        if (authenticationContext->state().testAndSwap(
+                State::e_AUTHENTICATING,
+                State::e_AUTHENTICATED) != State::e_AUTHENTICATING) {
+            errorStream << "Failed to set authentication state for '"
+                        << channel->peerUri()
+                        << "' to 'e_AUTHENTICATED' from 'e_AUTHENTICATING'";
+            error = errorStream.str();
+            return rc_AUTHENTICATION_STATE_INCORRECT;
+        }
+
+        // Schedule authentication timeout
+        if (result->lifetimeMs().has_value()) {
+            bslmt::LockGuard<bslmt::Mutex> lockGuard(
+                &authenticationContext->timeoutHandleMutex());  // MUTEX LOCKED
+
+            if (authenticationContext->timeoutHandle()) {
+                d_scheduler.cancelEvent(
+                    &authenticationContext->timeoutHandle());
+            }
+
+            d_scheduler.scheduleEvent(
+                &authenticationContext->timeoutHandle(),
+                bsls::TimeInterval(bmqsys::Time::nowMonotonicClock())
+                    .addMilliseconds(result->lifetimeMs().value()),
+                bdlf::BindUtil::bind(
+                    &Authenticator::reauthenticateErrorOrTimeout,
+                    this,                     // authenticator
+                    -1,                       // errorCode
+                    "authenticationTimeout",  // errorName
+                    channel                   // channel
+                    ));
+        }
+    }
+
+    return rc_SUCCESS;
+}
+
 // CREATORS
-Authenticator::Authenticator(BlobSpPool*       blobSpPool,
-                             bslma::Allocator* allocator)
-: d_allocator_p(allocator)
+Authenticator::Authenticator(
+    mqbauthn::AuthenticationController* authnController,
+    BlobSpPool*                         blobSpPool,
+    bslma::Allocator*                   allocator)
+: d_authnController_p(authnController)
+, d_threadPool(bmqsys::ThreadUtil::defaultAttributes(),
+               0,                                            // min threads
+               100,                                          // max threads
```

Comment:
This is a lot of threads.  How did we pick this number?

Maybe we need this many, but it should probably be configurable.